src/rebar_dialyzer_format.erl

```diff
@@ -59,6 +57,16 @@ format_warning_(Opts, Warning = {_Tag, {SrcFile, Line}, Msg}, {_LastFile, Acc})
             {SrcFile, [dialyzer:format_warning(Warning, fullpath) | Acc]}
     end.
 
+base_and_dir([] = _SrcFile, _Opts) ->
```

Comment:
I moved the logic closer to the source, which will produce
```erlang
/<path unknown>
   0: Unknown type b:b/0
```
(and seemingly is not affecting the tests)

Reply:
If you prefer not to have that "/" I can add an exception to that `filename:join`, e.g.
```erlang
filename_join("" = _Path, Base) ->
    Base;
filename_join(Path, Base) ->
    filename:join(Path, Base).
```

Reply:
I would probably move it into a different function clause entirely. Almost everything related to it in there is about managing the paths, which we don't require in this case.

Something like:

```erlang
format_warning_(Opts, Warning = {_Tag, {[], Line=0}, Msg}, {_LastFile, Acc}) ->
    SrcFile = "<path unknown>",
    try
        String = message_to_string(Msg),
        F = fmt("~!__~ts", [SrcFile]),
        {SrcFile, [lists:flatten(fmt("~n~ts~n~!c~4w~!!: ~ts", [F, Line, String])) | Acc]}
    catch
        ?WITH_STACKTRACE(Error, Reason, Stacktrace)
            ?DEBUG("Failed to pretty format warning: ~p:~p~n~p",
                   [Error, Reason, Stacktrace]),
            {SrcFile, [dialyzer:format_warning(Warning, fullpath) | Acc]}
    end.
format_warning_(Opts, Warning = {_Tag, {SrcFile, Line}, Msg}, {_LastFile, Acc}) ->
    try
        File = rebar_dir:format_source_file_name(SrcFile, Opts),
        Base = filename:basename(File),
        Dir = filename:dirname(File),
        Root = filename:rootname(Base),
        Ext = filename:extension(Base),
        Path = re:replace(Dir, "^.*/_build/", "_build/", [{return, list}, unicode]),
        Base1 = fmt("~!_c~ts~!!~!__~ts", [Root, Ext]),
        F = fmt("~!__~ts", [filename:join(Path, Base1)]),
        String = message_to_string(Msg),
        {SrcFile, [lists:flatten(fmt("~n~ts~n~!c~4w~!!: ~ts", [F, Line, String])) | Acc]}
    catch
        ?WITH_STACKTRACE(Error, Reason, Stacktrace)
            ?DEBUG("Failed to pretty format warning: ~p:~p~n~p",
                   [Error, Reason, Stacktrace]),
            {SrcFile, [dialyzer:format_warning(Warning, fullpath) | Acc]}
    end.
```

There's no reason to go through exactly the same assumption that there is a base and a directory when the values don't exist.

Reply:
Yup, I went the lazy way :D I didn't want to learn how to `fmt`. Let me try it.